File: 2020/24.py
import itertools

movs = {
    "e":  lambda p: (p[0] + 1, p[1] - 1, p[2]),
    "w":  lambda p: (p[0] - 1, p[1] + 1, p[2]),
    "se": lambda p: (p[0], p[1] - 1, p[2] + 1),
    "sw": lambda p: (p[0] - 1, p[1], p[2] + 1),
    "nw": lambda p: (p[0], p[1] + 1, p[2] - 1),
    "ne": lambda p: (p[0] + 1, p[1], p[2] - 1),
}

def neigh(p):
    for f in movs.values():
        yield f(p)

def main(inp):
    alive = set()
    for l in inp.split('\n'):
        p = (0, 0, 0)
        i = 0
        while i < len(l):
            mov = l[i]
            i += 1
            if mov in "sn":
                mov += l[i]
                i += 1
            p = movs[mov](p)
        (alive.add if p not in alive else alive.discard)(p)
    print(len(alive))
    for t in range(100):
        def good(p):
            cnt = sum(map(alive.__contains__, neigh(p)))
            return cnt == 2 or (p in alive and cnt == 1)
        totest = itertools.chain.from_iterable(map(neigh, alive))
        alive = {p for p in totest if good(p)}
    print(len(alive))

from aocd import get_data
main(get_data(day=24))

File: 2020/24.jl
movs = Dict(
    "e"  => (x, y, z) -> (x + 1, y - 1, z),
    "w"  => (x, y, z) -> (x - 1, y + 1, z),
    "se" => (x, y, z) -> (x, y - 1, z + 1),
    "sw" => (x, y, z) -> (x - 1, y, z + 1),
    "nw" => (x, y, z) -> (x, y + 1, z - 1),
    "ne" => (x, y, z) -> (x + 1, y, z - 1),
)

function neigh(p::Tuple{Int, Int, Int})
    c = Tuple{Int, Int, Int}[]
    for f in values(movs)
        push!(c, f(p...))
    end
    return c
end

function main(inp)
    alive = Set()
    for l in inp
        p = (0, 0, 0)
        i = 1
        while i <= length(l)
            if l[i] == 's' || l[i] == 'n'
                mov = l[i:i+1]
                i += 2
            else
                mov = l[i:i]
                i += 1
            end
            p = movs[mov](p...)
        end
        (p ∈ alive ? pop! : push!)(alive, p)
    end
    println(length(alive))
    for t in 1:100
        function good(p)
            cnt = length([x for x in neigh(p) if x ∈ alive])
            return cnt == 2 || (p ∈ alive && cnt == 1)
        end
        totest = Iterators.flatten(neigh.(alive))
        alive = Set([p for p in totest if good(p)])
    end
    println(length(alive))
end

main(readlines("24.in"))
